rmf_demo_plugins/rmf_gazebo_plugins/src/TeleportDispenser.cpp

```diff
@@ -0,0 +1,273 @@
+  {
+    const auto r_model = _world->ModelByName(rs.name);
+    if (r_model)
+      robot_models.push_back(r_model);
+  }
+
+  std::string nearest_robot_model_name;
+  if (!find_nearest_non_static_model_name(
+      robot_models, nearest_robot_model_name))
+  {
+    RCLCPP_WARN(_dispenser_common->ros_node->get_logger(),
+      "No near robots of fleet [%s] found.", fleet_name.c_str());
+    return;
+  }
+  _item_model->PlaceOnEntity(nearest_robot_model_name);
+  _dispenser_common->dispenser_filled = false;
+}
+
+// Searches vicinity of Dispenser for closest valid item. If found, _item_model is set to the newly found item
+void TeleportDispenserPlugin::fill_dispenser()
+{
+  auto model_list = _world->Models();
+  double nearest_dist = 1.0;
+  const auto dispenser_pos = _model->WorldPose().Pos();
+  for (const auto& m : model_list)
+  {
+    if (!m || m->IsStatic() || m->GetName() == _model->GetName())
+      continue;
+
+    const double dist = m->WorldPose().Pos().Distance(dispenser_pos);
+    if (dist < nearest_dist &&
+      _dispenser_vicinity_box.Intersects(m->BoundingBox()))
+    {
+      _item_model = m;
+      nearest_dist = dist;
+      _dispenser_common->dispenser_filled = true;
+    }
+  }
+}
+
+void TeleportDispenserPlugin::create_dispenser_bounding_box()
+{
+  // Find the dispenser item model, maximum distance 1 meter
+  const auto dispenser_pos = _model->WorldPose().Pos();
+  ignition::math::Vector3d corner_1 = dispenser_pos;
+  corner_1.X(dispenser_pos.X() - 0.05);
+  corner_1.Y(dispenser_pos.Y() - 0.05);
+  corner_1.Z(dispenser_pos.Z() - 0.05);
+  ignition::math::Vector3d corner_2 = dispenser_pos;
+  corner_2.X(dispenser_pos.X() + 0.05);
+  corner_2.Y(dispenser_pos.Y() + 0.05);
+  corner_2.Z(dispenser_pos.Z() + 0.05);
+  #if GAZEBO_MAJOR_VERSION <= 9
+  _dispenser_vicinity_box = ignition::math::Box(corner_1, corner_2);
+  #else
+  _dispenser_vicinity_box =
+    ignition::math::AxisAlignedBox(corner_1, corner_2);
+  #endif
+}
+
+void TeleportDispenserPlugin::on_update()
+{
+  if (!_load_complete)
+    return;
+
+  _dispenser_common->sim_time = _world->SimTime().Double();
+
+  // `_dispense` is set to true if the dispenser plugin node has received a valid DispenserRequest
+  if (_dispenser_common->dispense)
+  {
+    _dispenser_common->send_dispenser_response(DispenserResult::ACKNOWLEDGED);
+
+    if (_dispenser_common->dispenser_filled)
+    {
+      RCLCPP_INFO(_dispenser_common->ros_node->get_logger(),
+        "Dispensing item");
+      dispense_on_nearest_robot(_dispenser_common->latest.transporter_type);
+
+      _dispenser_common->send_dispenser_response(DispenserResult::SUCCESS);
+    }
+    else
+    {
+      RCLCPP_WARN(_dispenser_common->ros_node->get_logger(),
+        "No item to dispense: [%s]", _dispenser_common->latest.request_guid);
+      _dispenser_common->send_dispenser_response(DispenserResult::FAILED);
+    }
+    _dispenser_common->dispense = false;
+  }
```

Comment:
I considered this earlier, but it involves making a call back to the `dispense_on_nearest_robot` function, and I wasn't sure if calling functions from the individual classes inside the common class would make things too complex. Also, since both `dispense_on_nearest_robot` functions take different arguments, I _think_ this may be possible using `std::bind` and `std::function`?

Either way as you said the savings from future maintenance and updates probably outweighs those concerns. Will look at moving it!